### PR TITLE
[CIS-578] Support Pinned Messages

### PR DESCRIPTION
**Pull request by nuno-vieira**
# In this PR

- Add pin support when creating new messages
- Add pin and unpin support for existing messages
- Add support for getting latest 10 pinnedMessages

### CODE REVIEW

File: Sources/StreamChat/Query/ChannelListQuery.swift [hunk 1]
```diff
@@ -74,6 +74,9 @@ public extension FilterKey where Scope: AnyChannelListFilterScope {
 
     /// A filter key for matching the `memberCount` value.
     static var memberCount: FilterKey<Scope, Int> { "member_count" }
+
+    /// A filter key for matching the `pinned` value.
+    static var pinned: FilterKey<Scope, Bool> { "pinned" }
```
Comment (VojtaStavik) [inline Sources/StreamChat/Query/ChannelListQuery.swift:79]: As talked in Slack, this is probably not correct

File: Sources/StreamChat/WebSocketClient/EventMiddlewares/ChannelPinnedMessagesUpdaterMiddleware.swift [hunk 2]
```diff
@@ -3,3 +3,83 @@
...
+                event: event,
+                cid: cid,
+                messageId: messageId,
+                handle: { (channel, message) in
+                    if message.pinned {
+                        channel.pinnedMessages.insert(message)
+                    }
+                },
+                completion: completion
+            )
```
Comment (VojtaStavik) [inline Sources/StreamChat/WebSocketClient/EventMiddlewares/ChannelPinnedMessagesUpdaterMiddleware.swift:54]: I think ideally this should be done in the payload saving code.

File: Sources/StreamChat/WebSocketClient/EventMiddlewares/ChannelPinnedMessagesUpdaterMiddleware.swift [hunk 3]
```diff
@@ -3,3 +3,83 @@
...
+                event: event,
+                cid: cid,
+                messageId: messageId,
+                handle: { (channel, message) in
+                    if message.pinned {
+                        channel.pinnedMessages.remove(message)
+                    }
+                },
+                completion: completion
+            )
```
Comment (VojtaStavik) [inline Sources/StreamChat/WebSocketClient/EventMiddlewares/ChannelPinnedMessagesUpdaterMiddleware.swift:40]: This is probably not needed. Ideally, we should approach this as any other message -> the message is still visible, but it's marked deleted.

**Review by VojtaStavik — COMMENTED** [review]
I really like this PR 👍 I'm not sure we need `ChannelPinnedMessagesUpdaterMiddleware`, seems to me like this should rather be handled by the serialization code. Let's talk in Slack

File: Sources/StreamChat/Controllers/ChannelController/ChannelController.swift [hunk 4]
```diff
@@ -944,6 +969,13 @@ public enum ListOrdering {
     case bottomToTop
 }
 
+// Describes the pin expiration
+public enum Pin {
+    case noExpiration
+    case expirationDate(Date)
+    case expirationTime(TimeInterval)
+}
+
```
[the remark this answers is not in the document]
Comment (nuno-vieira) [reply Sources/StreamChat/Controllers/ChannelController/ChannelController.swift:978]: I think this simplifies a bit the API by adding this abstraction. Some customers might not know How to use the `Date` API. So with this abstraction, everyone will understand.

Also, we need to provide a way to `pin` and `unpin` a message. We can't just provide a date. Because if `nil` would mean `noExpiration` then how would you not pin a message, or unpin a message? This would require us to have a different property, like `pinned = true`. This is actually what the other platforms are doing but I think it is bad because it makes it possible to make a mistake and write `pinned = false` and `expirationDate = Date()`, which is a wrong state.

With this enum, we can describe 4 states. 
- No pin (`Pin? = nil`)
- Pin without expiration
- Pin with an expiration date
- Pin with the amount of time.

File: Sources/StreamChat/Models/Message.swift [hunk 5]
```diff
@@ -162,6 +171,18 @@ public enum MessageType: String, Codable {
...
+// The pinning information of a message.
+public struct _MessagePinDetails<ExtraData: ExtraDataTypes> {
+    /// Date when the message got pinned
+    var pinnedAt: Date?
+
+    /// The user that pinned the message
+    var pinnedBy: _ChatUser<ExtraData.User>?
+
+    /// Date when the message pin expires. An nil value means that message does not expire
+    var expiresAt: Date?
```
[the remark this answers is not in the document]
Comment (nuno-vieira) [reply Sources/StreamChat/Models/Message.swift:183]: The backend says they are optional. But I guess that if the message is pinned, these properties can't be optional. I played safe here and let them optional anyways. 

Other approaches:
1. In the parsing I can assume that if I can't parse these values, the message is not pinned, even if it says "pinned = true".
2. If we find "pinned = true" we parse those values and make sure they exist, if not, it fails the parsing and crashes.

Maybe option 1. is the best approach?

File: Sources/StreamChat/APIClient/Endpoints/Payloads/MessagePayloads.swift [hunk 6]
```diff
@@ -64,6 +68,11 @@ class MessagePayload<ExtraData: ExtraDataTypes>: Decodable {
     let attachments: [AttachmentPayload]
     let isSilent: Bool
 
+    var pinned: Bool
```
[the remark this answers is not in the document]
Comment (nuno-vieira) [reply Sources/StreamChat/APIClient/Endpoints/Payloads/MessagePayloads.swift:71]: It possible according to the Backend API, but I would say it would be a bug. Better to confirm this tho.

File: Sources/StreamChat/Database/DTOs/MessageDTO.swift [hunk 7]
```diff
@@ -191,12 +196,35 @@ extension MessageDTO {
...
+            pinExpires = pinnedDate.addingTimeInterval(time)
+        }
+    }
+
+    func unpin() {
+        pinned = false
+        pinnedAt = nil
+        pinnedBy = nil
+        pinExpires = nil
+    }
```
[the remark this answers is not in the document]
Comment (nuno-vieira) [reply Sources/StreamChat/Database/DTOs/MessageDTO.swift:220]: Thanks, will do 👍

File: Sources/StreamChat/WebSocketClient/EventMiddlewares/ChannelPinnedMessagesUpdaterMiddleware.swift (hunk 2 above)
Comment (nuno-vieira) [reply Sources/StreamChat/WebSocketClient/EventMiddlewares/ChannelPinnedMessagesUpdaterMiddleware.swift:54]: Yes makes totally sense 👍

File: Tests/StreamChatTests/MockEnpointResponses/Events/Notification/NotificationMessageNew.json [hunk 8]
```diff
@@ -148,7 +148,11 @@
         "mentioned_users" : [
             
         ],
-        "html" : "<p>Hi<\/p>\n"
+        "html" : "<p>Hi<\/p>\n",
+        "pinned": false,
+        "pinned_at": null,
+        "pin_expires": null,
+        "pinned_by": null
```
Comment (VojtaStavik) [inline Tests/StreamChatTests/MockEnpointResponses/Events/Notification/NotificationMessageNew.json:155]: Thank you for updating all payloads! nit: Ideally these changes would be in a separate commit so it's easier to review.

File: Tests/StreamChatTests/MockEnpointResponses/Channel.json [hunk 9]
```diff
@@ -1191,6 +1191,59 @@
...
+        ],
+        "updated_at" : "2020-06-09T08:10:40.800912Z",
+        "reply_count" : 0,
+        "user" : {
+          "id" : "broken-waterfall-5",
+          "banned" : false,
+          "unread_channels" : 0,
+          "extraData" : {
+            "name" : "Tester"
+          },
```
Comment (VojtaStavik) [inline Tests/StreamChatTests/MockEnpointResponses/Channel.json:1224]: Where did you get these data from? This is not the way we save `extraData` :)

File: Sources/StreamChat/Controllers/ChannelController/ChannelController.swift [hunk 10]
```diff
@@ -941,6 +944,21 @@ public enum ListOrdering {
     case bottomToTop
 }
 
+// Describes the pinning expiration
+public struct Pin: Equatable {
```
Comment (VojtaStavik) [inline Sources/StreamChat/Controllers/ChannelController/ChannelController.swift:973]: Can we rename this to `Pinning`? Also, it would be great to add documentation for public APIs

File: Sources/StreamChat/Controllers/ChannelController/ChannelController.swift [hunk 11]
```diff
@@ -735,6 +735,31 @@ public extension _ChatChannelController {
...
+    ///   - messageId: The message identifier.
+    ///   - completion: A completion block with an error if the request was failed.
+    func unpinMessage(messageId: MessageId, completion: ((Error?) -> Void)? = nil) {
+        updater.unpinMessage(messageId: messageId) { result in
+            self.callback {
+                completion?(result)
+            }
+        }
+    }
+
```
Comment (VojtaStavik) [inline Sources/StreamChat/Controllers/ChannelController/ChannelController.swift:762]: This modifies the messages so shouldn't this rather be on `MessageController`?

File: Sources/StreamChat/Database/DTOs/MessageDTO.swift [hunk 12]
```diff
@@ -375,6 +375,24 @@ extension NSManagedObjectContext: MessageDatabaseSession {
     func delete(message: MessageDTO) {
         delete(message)
     }
+
+    func pin(message: MessageDTO, _ pin: Pin) throws {
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/MessageDTO.swift:403]: Why is it `_ pin` and not `pin`? 🤷‍♂️ :) 

I think with the adjusted naming it will work nicely `func pin(message: MessageDTO, pinning: MessagePinning)`

File: Sources/StreamChat/Controllers/ChannelController/ChannelController_Tests.swift [hunk 13]
```diff
@@ -594,8 +594,8 @@ class ChannelController_Tests: StressTestCase {
         }
         XCTAssertNil(error)
         let channel: ChatChannel = client.databaseContainer.viewContext.channel(cid: channelId)!.asModel()
-        assert(channel.latestMessages.count == 1)
-        let message: ChatMessage = channel.latestMessages.first!
+        XCTAssertEqual(channel.latestMessages.count, 1)
+        let message: ChatMessage = try XCTUnwrap(channel.latestMessages.first)
```
Comment (VojtaStavik) [inline Sources/StreamChat/Controllers/ChannelController/ChannelController_Tests.swift:598]: The semantic difference between `XCTAssert` and `assert` is that by using `assert` you're saying - this is how it should be and it's not the thing we're testing here. However, I do understand the original implementation wasn't the best because it started crashing once you introduced pinned messages. Thanks for improving 👍

File: Sources/StreamChat/Database/DTOs/ChannelDTO.swift [hunk 14]
```diff
@@ -20,6 +20,10 @@ class ChannelDTO: NSManagedObject {
     @NSManaged var updatedAt: Date
     @NSManaged var lastMessageAt: Date?
 
+    // The oldest message in the current batch of messages. This property only lives locally, and it is useful
+    // to filter out older pinned messages that do not belong to the current batch of messages in the channel.
+    @NSManaged var oldestMessageAt: Date
+
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/ChannelDTO.swift:26]: `// The oldest message in the current batch of messages.` is not correct, it's the oldest message of the channel we have locally coming from a regular channel query.

File: Sources/StreamChat/Database/DTOs/ChannelDTO.swift [hunk 15]
```diff
@@ -124,6 +128,10 @@ extension NSManagedObjectContext {
         dto.lastMessageAt = payload.lastMessageAt
         dto.memberCount = Int64(clamping: payload.memberCount)
 
+        if let oldestMessageAt = dto.messages.map(\.createdAt).min() {
+            dto.oldestMessageAt = oldestMessageAt
+        }
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/ChannelDTO.swift:133]: We should override only if `dto.oldestMessageAt < oldestMessageAt`. This would be good to test, too, btw.

File: Sources/StreamChat/Database/DTOs/ChannelDTO_Tests.swift [hunk 16]
```diff
@@ -7,6 +7,8 @@ import XCTest
 
 class ChannelDTO_Tests: XCTestCase {
     var database: DatabaseContainer!
+
+    static let channelCreatedDate = "2019-12-12T15:33:46.488935Z".toDate()
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/ChannelDTO_Tests.swift:11]: To make this more fun, you can make this `.unique`, too. It will be unique for the duration of the tests, but it will change with every test suite run to provide additional variability in the tests.

File: Sources/StreamChat/Database/DTOs/ChannelDTO_Tests.swift [hunk 17]
```diff
@@ -209,6 +211,72 @@ class ChannelDTO_Tests: XCTestCase {
         XCTAssertEqual(channel?.latestMessages.count, 10)
     }
 
+    func test_channelPayload_pinnedMessagesOlderThanOldestMessageAtAreIgnored() throws {
+        // Save a channel payload with 100 messages
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/ChannelDTO_Tests.swift:215]: that `100` there is probably a copy/paste mistake? Also, why even use 20? Wouldn't just 1+1 be better and more obvious? :)

File: Sources/StreamChat/Database/DTOs/ChannelDTO_Tests.swift [hunk 18]
```diff
@@ -209,6 +211,72 @@ class ChannelDTO_Tests: XCTestCase {
...
+        let channelId: ChannelId = .unique
+        let oldPinnedMessage: MessagePayload<NoExtraData> = MessagePayload(
+            id: .unique,
+            type: .regular,
+            user: dummyUser,
+            createdAt: "2018-12-12T15:33:46.488935Z".toDate(),
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/ChannelDTO_Tests.swift:221]: nit: Why not just use some more readable version of the date? The actual value doesn't matter, it just has to be older than a specific date.

File: Sources/StreamChat/Database/DTOs/MessageDTO.swift [hunk 19]
```diff
@@ -117,11 +117,18 @@ class MessageDTO: NSManagedObject {
             .init(format: "createdAt > channel.truncatedAt")
         ])
 
+        // Some pinned messages might be in the local database, but should not be fetched
+        // if they are not in the current batch of messages.
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/MessageDTO.swift:121]: Same here - the "current batch of messages" is imo a little bit confusing wording.

File: Sources/StreamChat/Database/DTOs/MessageDTO.swift [hunk 20]
```diff
@@ -331,15 +338,32 @@ extension NSManagedObjectContext: MessageDatabaseSession {
...
+        var channelDTO: ChannelDTO?
+
         if let channelPayload = payload.channel {
-            let channelDTO = try saveChannel(payload: channelPayload, query: nil)
-            dto.channel = channelDTO
+            channelDTO = try saveChannel(payload: channelPayload, query: nil)
         } else if let cid = cid {
-            let channelDTO = ChannelDTO.loadOrCreate(cid: cid, context: self)
-            dto.channel = channelDTO
+            channelDTO = ChannelDTO.loadOrCreate(cid: cid, context: self)
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/MessageDTO.swift:346]: nit: We talked about the fact that this should be just `load`

File: Sources/StreamChat/Database/DTOs/MessageDTO.swift [hunk 21]
```diff
@@ -331,15 +338,32 @@ extension NSManagedObjectContext: MessageDatabaseSession {
...
+            } else {
+                // If message is not pinned, or was unpinned, should be removed from pinnedMessages
+                channelDTO.pinnedMessages.remove(dto)
+                // If the current message dto is pinned, should not be included to oldestMessagetAt
+                messagesCreatedAt.append(dto.createdAt)
+            }
+            if let oldestMessageAt = messagesCreatedAt.min() {
+                channelDTO.oldestMessageAt = oldestMessageAt
+            }
+            dto.channel = channelDTO
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/MessageDTO.swift:365]: I'm confused by this. I think we shouldn't touch the `channelDTO.oldestMessageAt` here at all. That's the main point of it - to update it only if it comes from the regular channel query. Implemented like this, if a message that's not presented locally is updated, it will adjust `channelDTO.oldestMessageAt` and we have exactly the problem we're trying to work around. What we do in `ChannelDTO` is enough IMO.

File: Sources/StreamChat/Database/DTOs/MessageDTO_Tests.swift [hunk 22]
```diff
@@ -242,7 +242,59 @@ class MessageDTO_Tests: XCTestCase {
...
+                // Save the message
+                let messageDTO = try! session.saveMessage(payload: payload, for: channelId)
+                completion((channelDTO, messageDTO))
+            }
+        }
+
+        XCTAssertEqual(channelDTO.oldestMessageAt, payload.createdAt)
+        XCTAssertFalse(channelDTO.pinnedMessages.contains(messageDTO))
+    }
+
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/MessageDTO_Tests.swift:297]: Yep, I think this is actually behavior we don't want

**Review by VojtaStavik — CHANGES_REQUESTED** [review]
This PR is full of interesting and complex solutions. Let's talk in Slack about the details.

File: Sources/StreamChat/APIClient/Endpoints/Payloads/ChannelListPayload_Tests.swift [hunk 23]
```diff
@@ -49,6 +53,7 @@ class ChannelPayload_Tests: XCTestCase {
         XCTAssert(firstMessage.reactionScores.isEmpty)
         XCTAssertEqual(firstMessage.replyCount, 0)
         XCTAssertFalse(firstMessage.isSilent)
+        XCTAssertEqual(firstPinnedMessage.id, "broken-waterfall-5-7aede36b-b89f-4f45-baff-c40c7c1875d9")
```
[the remark this answers is not in the document]
Comment (nuno-vieira) [reply Sources/StreamChat/APIClient/Endpoints/Payloads/ChannelListPayload_Tests.swift:56]: 🤦 An assertNotNil here is enough 😄

File: Tests/StreamChatTests/MockEnpointResponses/Channel.json (hunk 9 above)
Comment (nuno-vieira) [reply Tests/StreamChatTests/MockEnpointResponses/Channel.json:1224]: I only changed data related to Pin information, so this extradata was already in this json 🤔

File: Sources/StreamChat/Controllers/ChannelController/ChannelController.swift (hunk 10 above)
Comment (nuno-vieira) [reply Sources/StreamChat/Controllers/ChannelController/ChannelController.swift:973]: Yes I added the docs to the functions, but forgot to add documentation directly to Pin 👍

File: Sources/StreamChat/Controllers/ChannelController/ChannelController.swift (hunk 11 above)
Comment (nuno-vieira) [reply Sources/StreamChat/Controllers/ChannelController/ChannelController.swift:762]: In the documentation all other platforms pin and unpin the message in the channelController. So this makes it consistent with the other platforms, and also kinda makes sense, because we pinning a message in a channel, and not pinning a message in all channels, don't know if this last sentence made sense 😄

File: Sources/StreamChat/Database/DTOs/MessageDTO.swift (hunk 12 above)
Comment (nuno-vieira) [reply Sources/StreamChat/Database/DTOs/MessageDTO.swift:403]: Yeah the hidden labelling was so it was not so repetitive. `pin(message: someMessage, .noExpirationDate)` instead of repeating `pin` again. But `MessagePinning` sounds good to me 👍

File: Sources/StreamChat/Controllers/ChannelController/ChannelController_Tests.swift (hunk 13 above)
Comment (nuno-vieira) [reply Sources/StreamChat/Controllers/ChannelController/ChannelController_Tests.swift:598]: Yes that was exactly why I changed this, but I understood why it was there in the first place. But it was kinda annoying to always crash when was trying to debug, so that is why I changed 👍

File: Sources/StreamChat/Database/DTOs/ChannelDTO.swift (hunk 14 above)
Comment (nuno-vieira) [reply Sources/StreamChat/Database/DTOs/ChannelDTO.swift:26]: Makes sense, thanks 👍

File: Sources/StreamChat/Database/DTOs/ChannelDTO.swift (hunk 15 above)
Comment (nuno-vieira) [reply Sources/StreamChat/Database/DTOs/ChannelDTO.swift:133]: Why is that check necessary? 🤔

File: Sources/StreamChat/Database/DTOs/ChannelDTO_Tests.swift (hunk 17 above)
Comment (nuno-vieira) [reply Sources/StreamChat/Database/DTOs/ChannelDTO_Tests.swift:215]: Yes copy paste issue 😄 Yeah you right, 1+1 will be easier to understand what we are testing here, will change it 👍

File: Tests/StreamChatTests/MockEnpointResponses/Channel.json (hunk 9 above)
Comment (nuno-vieira) [reply Tests/StreamChatTests/MockEnpointResponses/Channel.json:1224]: I extracted a message, from another payload, so that payload has the extraData like this, probably it is a old payload. I'm going to remove it from here

Comment (nuno-vieira) [reply Tests/StreamChatTests/MockEnpointResponses/Channel.json:1224]: We have it in a lot of places tho :D Do you want me to remove them all?
image

Comment (VojtaStavik) [reply Tests/StreamChatTests/MockEnpointResponses/Channel.json:1224]: @nuno-vieira if you can remove it please do, it will only cause confusion in the future.

File: Sources/StreamChat/APIClient/Endpoints/Payloads/ChannelListPayload_Tests.swift [hunk 24]
```diff
@@ -49,6 +53,7 @@ class ChannelPayload_Tests: XCTestCase {
         XCTAssert(firstMessage.reactionScores.isEmpty)
         XCTAssertEqual(firstMessage.replyCount, 0)
         XCTAssertFalse(firstMessage.isSilent)
+        XCTAssertNotNil(firstPinnedMessage)
```
Comment (VojtaStavik) [inline Sources/StreamChat/APIClient/Endpoints/Payloads/ChannelListPayload_Tests.swift:56]: Maybe we can replace
```swift
XCTAssertEqual(payload.pinnedMessages.count, 1)
let firstPinnedMessage = payload.pinnedMessages
    .first(where: { $0.id == "broken-waterfall-5-7aede36b-b89f-4f45-baff-c40c7c1875d9" })
XCTAssertNotNil(firstPinnedMessage)
```

with just

```swift
XCTAssertEqual(
    payload.pinnedMessages.map { $0.id }, 
    ["broken-waterfall-5-7aede36b-b89f-4f45-baff-c40c7c1875d9"]
)
```
??

File: Sources/StreamChat/Database/DTOs/ChannelDTO.swift [hunk 25]
```diff
@@ -153,6 +154,10 @@ extension NSManagedObjectContext {
         let dto = try saveChannel(payload: payload.channel, query: query)
         
         try payload.messages.forEach { _ = try saveMessage(payload: $0, for: payload.channel.cid) }
+        try payload.pinnedMessages.forEach {
+            let message = try saveMessage(payload: $0, for: payload.channel.cid)
+            dto.pinnedMessages.insert(message)
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/ChannelDTO.swift:173]: Is this needed? This is already happening inside `saveMessage`

File: Sources/StreamChat/Models/Channel.swift [hunk 26]
```diff
@@ -96,6 +96,13 @@ public struct _ChatChannel<ExtraData: ExtraDataTypes> {
     /// and using a `ChatChannelController` for this channel id.
     ///
     public let latestMessages: [_ChatMessage<ExtraData>]
+
+    /// Pinned messages present on the channel.
+    ///
+    /// This field contains only the pinned messages of the channel. You can get all existing messages in the channel by creating
+    /// and using a `ChatChannelController` for this channel id.
+    ///
```
Comment (VojtaStavik) [inline Sources/StreamChat/Models/Channel.swift:104]: Now when we have it, it would make sense to actually use the `@Cached` mechanism for `pinnedMessages`, too.

File: Sources/StreamChat/Database/DTOs/ChannelDTO_Tests.swift [hunk 27]
```diff
@@ -166,6 +175,19 @@ class ChannelDTO_Tests: XCTestCase {
...
+        // Save a channel payload with 100 messages
+        let channelId: ChannelId = .unique
+        let payload = dummyPayload(with: channelId, numberOfMessages: 100)
+
+        try database.writeSynchronously { session in
+            try session.saveChannel(payload: payload)
+        }
+
+        let channel: ChatChannel? = database.viewContext.channel(cid: channelId)?.asModel()
+        XCTAssertEqual(channel?.pinnedMessages.count, 1)
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/ChannelDTO_Tests.swift:188]: For the test here it's not clear why `pinnedMessages.count` should be 1

File: Sources/StreamChat/Database/DTOs/ChannelDTO_Tests.swift [hunk 28]
```diff
@@ -7,6 +7,8 @@ import XCTest
 
 class ChannelDTO_Tests: XCTestCase {
     var database: DatabaseContainer!
+
+    static let channelCreatedDate = Date.unique
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/ChannelDTO_Tests.swift:11]: Can you please add a comment explaining why this is needed and what problem it solves?

File: Sources/StreamChat/Database/DTOs/MessageDTO.swift [hunk 29]
```diff
@@ -326,15 +326,24 @@ extension NSManagedObjectContext: MessageDatabaseSession {
...
         }
+
+        if let channelDTO = channelDTO {
+            if dto.pinned {
+                channelDTO.pinnedMessages.insert(dto)
+            } else {
+                channelDTO.pinnedMessages.remove(dto)
+            }
+            dto.channel = channelDTO
+        }
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/MessageDTO.swift:358]: if `channelDTO` is `nil` we silently fail but in fact, it's a very serious issue.

File: Sources/StreamChat/Controllers/ChannelController/ChannelController.swift [hunk 30]
```diff
@@ -690,15 +690,17 @@ public extension _ChatChannelController {
     ///
     /// - Parameters:
     ///   - text: Text of the message.
-    ///   - extraData: Additional extra data of the message object.
+    ///   - pinning: Pins the new message. Nil if should not be pinned.
```
Comment (VojtaStavik) [inline Sources/StreamChat/Controllers/ChannelController/ChannelController.swift:693]: nit: `Nil` -> `` `nil` ``

File: Sources/StreamChat/Controllers/MessageController/MessageController.swift [hunk 31]
```diff
@@ -193,6 +193,7 @@ public extension _ChatMessageController {
     ///
     /// - Parameters:
     ///   - text: Text of the message.
+    ///   - pinning: Pins the new message. Nil if should not be pinned.
```
Comment (VojtaStavik) [inline Sources/StreamChat/Controllers/MessageController/MessageController.swift:196]: same here

File: Sources/StreamChat/Models/MessagePinning.swift [hunk 32]
```diff
@@ -0,0 +1,31 @@
...
+    public static func expirationDate(_ date: Date) -> Self {
+        .init(expirationDate: date)
+    }
+
+    /// Pins a message.
+    /// - Parameter time: The amount of seconds the message will be pinned.
+    public static func expirationTime(_ time: TimeInterval) -> Self {
+        .init(expirationDate: Date().addingTimeInterval(time))
+    }
+}
```
Comment (VojtaStavik) [inline Sources/StreamChat/Models/MessagePinning.swift:31]: Nice 🎉

**Review by VojtaStavik — APPROVED** [review]
Looks great 👍 Just some minor comments to test + the issue with overriding `oldestMessageAt` and it's good to 🚢 🎉

**Review by VojtaStavik — APPROVED** [review]
All right! Let's 🚢 this beauty 🎉

**Comment by VojtaStavik** [comment]
@nuno-vieira before merging, please add a CHANGELOG entry for this
